modin/core/dataframe/algebra/groupby.py

```diff
@@ -373,16 +413,201 @@ def try_filter_dict(agg_func, df):
+            by the `agg_func` keys (only makes sense when `agg_func` is a dictionary).
 
         Returns
         -------
-        Callable
+        callable(pandas.core.groupby.DataFrameGroupBy) -> pandas.DataFrame
             Aggregation function that can be safely applied to this particular partition.
         """
         if not isinstance(agg_func, dict):
             return agg_func
-        partition_dict = {k: v for k, v in agg_func.items() if k in df.columns}
-        return lambda grp: grp.agg(partition_dict)
+
+        grp_has_id_level = df.columns.names[0] == cls.ID_LEVEL_NAME
+        # The 'id' level prevents us from a lookup for the original
+        # partition's columns. So dropping the level.
+        partition_columns = frozenset(
+            df.columns.droplevel(0) if grp_has_id_level else df.columns
+        )
+
+        partition_dict = {k: v for k, v in agg_func.items() if k in partition_columns}
+        return cls._build_callable_for_dict(
+            partition_dict, preserve_aggregation_order, grp_has_id_level
+        )
+
+    @classmethod
+    def _build_callable_for_dict(
+        cls, agg_dict, preserve_aggregation_order=True, grp_has_id_level=False
+    ):
+        """
+        Build callable for an aggregation dictionary.
+
+        Parameters
+        ----------
+        agg_dict : dict
+            Aggregation dictionary.
+        preserve_aggregation_order : bool, default: True
+            Whether to manually restore the order of columns for the result specified
+            by the `agg_func` keys (only makes sense when `agg_func` is a dictionary).
+        grp_has_id_level : bool, default: False
+            Whether the frame we're grouping on has intermediate columns
+            (see ``cls.ID_LEVEL_NAME``).
+
+        Returns
+        -------
+        callable(pandas.core.groupby.DataFrameGroupBy) -> pandas.DataFrame
+        """
+        external_aggs = {}
+        native_aggs = {}
+
+        result_columns = []
+        for col, func, func_name, col_renaming_required in walk_aggregation_dict(
+            agg_dict
+        ):
+            # Filter dictionary
+            dict_to_add = (
+                external_aggs if cls.is_registered_implementation(func) else native_aggs
+            )
+
+            new_value = func if func_name is None else (func_name, func)
+            old_value = dict_to_add.get(col)
+
+            if old_value is not None:
+                ErrorMessage.catch_bugs_and_request_email(
+                    failure_condition=not isinstance(old_value, list),
+                    extra_log="Expected for all aggregation values to be a list when at least "
+                    + f"one column has multiple aggregations. Got: {old_value} {type(old_value)}",
+                )
+                old_value.append(new_value)
+            else:
+                dict_to_add[col] = [new_value] if col_renaming_required else new_value
```

Comment:
This is not obvious to me that if col_renaming_required is True, then we should wrap new_value in a list. Maybe it is worth adding a comment?

Reply:
Added the following in-code comment:
```python
                # Pandas knows that it has to modify the resulting columns if it meets
                # a function wrapped into a list. Renaming is required if either a new
                # column name was explicitly specified, or multiple functions were
                # specified per one column, or if any other column in the aggregation
                # is going to be renamed.
```